src/main/java/com/amazon/opendistroforelasticsearch/ad/ml/CheckpointDao.java

```
@@ -98,14 +203,135 @@ public void putModelCheckpoint(String modelId, String modelCheckpoint, ActionLis
+                new IndexRequest(indexName).id(modelId).source(source),
                 client::index,
                 ActionListener.wrap(r -> listener.onResponse(null), listener::onFailure)
             );
     }
 
+    /**
+     * Bulk writing model states prepared previously
+     */
+    public void flush() {
+        try {
+            // in case that other threads are doing bulk as well.
+            if (!lock.tryLock()) {
+                return;
+            }
+            if (requests.size() > 0 && bulkRateLimiter.tryAcquire()) {
+                final BulkRequest bulkRequest = new BulkRequest();
+                // at most 1000 index requests per bulk
+                for (int i = 0; i < maxBulkRequestSize; i++) {
+                    DocWriteRequest<?> req = requests.poll();
+                    if (req == null) {
+                        break;
+                    }
+
+                    bulkRequest.add(req);
+                }
+                if (indexUtil.doesCheckpointIndexExist()) {
+                    flush(bulkRequest);
+                } else {
+                    indexUtil.initCheckpointIndex(ActionListener.wrap(initResponse -> {
+                        if (initResponse.isAcknowledged()) {
+                            flush(bulkRequest);
+                        } else {
+                            throw new RuntimeException("Creating checkpoint with mappings call not acknowledged.");
+                        }
+                    }, exception -> {
+                        if (ExceptionsHelper.unwrapCause(exception) instanceof ResourceAlreadyExistsException) {
+                            // It is possible the index has been created while we sending the create request
+                            flush(bulkRequest);
+                        } else {
+                            logger.error(String.format("Unexpected error creating index %s", indexName), exception);
```

In which case we will call `write()` method of checkpointDao? Without knowing that, I am not sure whether it is okay to swallow the exception here.

---

we will call it whenever we need to save a checkpoint.  Any better option to not swallow it?

---

what if we throw the unexpected exception, will it break anything?

---

not completely sure without testing.  I guess the upstream will eventually swallow it.

---

i see. thanks.